tests/suites/test_suite_ecp.function

```diff
@@ -637,6 +637,15 @@ void ecp_read_binary( int id, data_t * buf, char * x, char * y, char * z,
         {
             TEST_ASSERT( mbedtls_mpi_cmp_mpi( &P.Y, &Y ) == 0 );
             TEST_ASSERT( mbedtls_mpi_cmp_mpi( &P.Z, &Z ) == 0 );
+
+            if( buf->x[0] == 0x04 )
+            {
+                /* re-encode in compressed format and test read again */
```

Comment:
I don't think this is the correct strategy. Reading compressed format is going to return `FEATURE_UNAVAILABLE` for curves where `p != 3 mod 4`, and there are a few, see https://github.com/Mbed-TLS/mbedtls/pull/521#discussion_r254614965 so we can't just assert that things work.

More generally speaking, we should have negative testing of reading points in compressed format: what if the data is the wrong length? What if the value of x is such that x^3 + a*x + b is not a square mod p (that is, it does not correspond to a point on the curve)?

So, I think the right strategy is t add more cases to the `.data` files, where we put compressed points in the test data, and encode the expected return value (plus the expected value of y in case of success, as a new parameter to this test function). For each curve we support, we need one case with correct data, and the expected return value will be 0 or `FEATURE_UNAVAILABLE` depending on the curve. Then for at least one curve which supports this, we need a few invalid inputs.

Reply:
I think reusing the same test data is a good strategy for positive test cases. But we do need negative test cases as well. And we need test cases for all curves ([even for uncompressed points](https://github.com/Mbed-TLS/mbedtls/issues/5940)).

Reply:
Right, we can can use a mixed approach: re-use test data for positive tests, and add new data for negative tests. We'll just need to adapt the code that re-uses test data for positive tests so that it expected `FEATURE_UNAVAILABLE` rather than `0` when p != 3 mod 4.

Reply:
> What if the value of x is such that x^3 + a*x + b is not a square mod p (that is, it does not correspond to a point on the curve)?

The current code in the library does not check that the x and y are points on the curve when the point is provided in uncompressed format.  Later use of the point, such as under `mbedtls_ecp_mul_mod()`, checks using `mbedtls_ecp_check_pubkey()`.  Therefore, if you would like me to add a call to `mbedtls_ecp_check_pubkey()` for compressed points, then it should probably be done for both uncompressed and compressed formats.  While doing so might be useful to catch invalid input earlier, this would add overhead to normal operation, and I can imagine that it might prevent other test cases from intentionally loading bad data from hex strings, though I have not specifically verified that.

I have added some test cases to tests/suites/test_suite_ecp.data and will push those shortly.  I think #5940 is a good idea but a bit too much scope creep to require here, and I hope that you'll agree.  The new function in the PR `mbedtls_ecp_sw_derive_y()` checks that prime p meets requirement p = 3 mod 4, and otherwise returns `MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE`, which matches the behavior of the existing code (which returns `MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE` for all points in compressed format).   I have added a test for that.

Please advise if you would like additional tests added.
:rocket:

Reply:
> The current code in the library does not check that the x and y are points on the curve when the point is provided in uncompressed format.

Good point. I had thought about that after commenting as well, and indeed we don't nee to do more validation when parsing compressed points than when parsing uncompressed points. However, we still need to document internally how the function behaves in this case.

Reply:
Prior behavior for compressed format was `MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE`, and the behavior is the same for prime p where p != 3 mod 4.  The difference is that the feature (to handle compressed point format) is available for prime p where p = 3 mod 4.

I believe things are already documented in `include/mbedtls/ecp.h` for `mbedtls_ecp_point_read_binary()`, including:
```
 * \note            This function does not check that the point actually
 *                  belongs to the given group, see mbedtls_ecp_check_pubkey()
 *                  for that.
```

Reply:
> If there is some clarification that you would like to see added, please be specific and I am happy to add it, or you may feel free to add a commit to this PR to do so. Thanks.

Sure, I think adding a comment on this internal function would be the right place, see https://github.com/Mbed-TLS/mbedtls/pull/6282#discussion_r985488107

(I marked this conversation as resolved, as the testing is OK, and the bit about non-squares is now tracked on its own.)